utilities/zookeeper/R/corrector-aardvark-II.R

```diff
@@ -0,0 +1,252 @@
+#' @param dump_locations character vector of locations to ignore
+#'
+#' @return A function that takes a list of covidcast signals as the only
+#'   argument
+#' @export
+#'
+#' @examples
+#' make_aardvark_corrector(default_state_params(window_size=21))
+make_aardvark_corrector <- function(
+  params = default_state_params(),
+  corrections_db_path = NULL,
+  dump_locations = c("as","gu","mp","vi")) {
+  
+  
+  aardvark_state_corrections <- function(df) {
+    if (class(df)[1] == "covidcast_signal") {
+      # in case there's only one signal
+      df <- list(df)
+    }
+    if (!is.null(dump_locations)) {
+      df <- purrr::map(df, ~.x %>%
+                         dplyr::filter(! geo_value %in% dump_locations))
+    }
+    params$to_correct <- TRUE # a key for deciding if we make corrections
+    in_names <- names(df[[1]])
+    all_signals <- df %>%
+      purrr::map_dfr(~.x %>%
+                       select(.data$data_source, .data$signal) %>%
+                       head(1))
+    params <- dplyr::left_join(all_signals, params)
+    if (all(is.na(params$to_correct))) {
+      warning(paste("No requested corrections data_source/signal pairs",
+                    "were actually present in the data. No corrections",
+                    "were implemented."))
+      return(df)
+    }
+    corrected <- list()
+    for (i in seq_along(df)) {
+      corrected[[i]] <- aardvark_state_corrections_single_signal(
+        df[[i]], params[i,])
+      df[[i]] <- corrected[[i]] %>%
+        mutate(value = .data$corrected) %>%
+        select(all_of(in_names)) %>%
+        ungroup()
+      df[[i]] <- covidcast::as.covidcast_signal(
+        df[[i]],
+        signal = df[[i]]$signal[1],
+        geo_type = params$geo_type[1])
+    }
+    
+    if (!is.null(corrections_db_path)) {
+      corrected_df <- bind_rows(corrected) %>%
+        select(.data$data_source, .data$signal, .data$geo_value,
+               .data$time_value,
+               .data$value, .data$corrected, .data$flag)
+      write_rds(corrected_df, file = corrections_db_path)
+    }
+    return(df)
+  }
+  
+  return(aardvark_state_corrections)
+}
+#' 
+# We add an argument "special_flags" that consists of lists for each states to the function "aardvark_state_corrections_single_signal"
+# There are four components in each state list: geo_value, signal. time_range, and max_lag. 
+#' @example 
+# list_WA <- list('wa', 'deaths_incidence_num',ymd(c("2020-12-16", "2020-12-17",
+#                                                     "2020-12-23", "2020-12-24",
+#                                                    "2020-12-29")), 14) 
+#
+# list_OH <- list('oh', 'deaths_incidence_num',seq(ymd("2021-02-12"),
+#                                                   ymd("2021-02-14"), length.out = 3), 60) 
+#
+# list_VA <- list('va','deaths_incidence_num',seq(ymd("2021-02-20"),ymd(time_value_flag_date),by=1), 60) 
+#
+# special_flags <- list(list_WA,list_OH,list_VA)
```

Comment:
I think this behaviour is pretty good. But let's try to improve it a bit. Rather than a list of lists, let's do a `tibble`. One column for the `geo_value` one for the `signal` and a third a "list column" for the times. This should allow the dates to be a vector of values or a function.

Reply:
Also, rather than doing it in a new file `corrector-aardvark-II.R`, just make changes to the current file. That way I can see what has changed in the PR.